Initial configuration for bag recording and foxglove bridge

hect95 opened the pull request:
This PR provides an initial configuration/setup for the ros2 bag recording and foxglove bridge from a docker container.

- Two types of scripts directories are introduced `host_tools` and `container_tools`. As it names implies `host_tools`provides scripts that are intended for host usage, i.e outside docker container. Moreover `container_tools`, offer scripts that are needed inside the docker container. The summary of the scripts states as follows:
  - `host_tools`:
    - `upload_docker_image_to_vehicle.sh`: Receive a docker image as input, to save it with `docker save` and send it over the vehicle uisng `scp`. Docker cmd are used for bash autocomplate, thus after doing "tab tab" method, the only docker images available in your system will be displayed
    - `install_tool.sh`: Install the upload docker script and autocompletion.
  - `container_tools`:
    - `foxglove_bridge.sh`: Runs foxglove_bridge launch file with some arguments
    - `record_rosbag.sh`: Runs `ros2 bag record` cmd to allow `mcap`  bag format. Optionally a `.txt` file can be passed as argument to define record-only topics.
   
 - Docker: The docker build implements dev and runtime build stages. Majority of the install relies on `base` stage. The default command for both sages are `bash` and `record_rosbag` respectively
 - `dev.sh` and `runtime.sh` scripts are provided to ease development and testing. By default BOTH scripts create a `rosbag` dir if it does not exists and mount it to the dockercontainer

hect95 committed:
Adding scripts to upload an image to vehicle

hect95 committed:
Improving vehilce upload script installation

hect95 committed:
Adding dockerfile and scripts

- Dockerfile runtime will run foxglove websocket bridge by default
- Adding dev.sh and runtime.sh for docker manipulation. Both scripts
	create a rosbags directory and mount it into container
- Adding .gitignore to exclude rosbags directory

hect95 committed:
Moving scripts that are meant to be used by host

hect95 committed:
Adding bag recording with mcap

- record_rosbag script is introduced to recrod mcap rosbag
	the script can recieve a .txt file  that contains
	the list of topics to be recorded. If no .txt file
	is provided the scritpp will use --all flag

- foxglove_bridge script is introduced to launch foxglove bridge
	with specific arguments

- config/sensor_topics.txt is added to define sensor ONLY topics
	to be recorded

- Dockerfile now copies config and scripts/container_tools dirs

- runtime stage now runs record_rosbag by default instead of
	foxglove bridge launch file

hect95 committed:
Adding pre-commit cfg file

hect95 committed:
Applying pre-commmit fixes

hect95 committed:
Installing more ros sensor msgs

hect95 committed:
Removing pre commit unused skip section

hect95 committed:
Attempting to fix hadolint ci break

hect95 committed:
Renaming host into host_tools

GreatAlexander commented on config/sensor_topics.txt:
Indeed, in future we could have a config file per use-case or per sensor, and then inc. whichever files correspond to an overal record configuration. But this is a good start for now.

hect95 replied:
This file is missing all the sensor topics, I will update it once I can see them on the vehicle

GreatAlexander commented on scripts/container_tools/foxglove_bridge:
```diff
@@ -0,0 +1,4 @@
+#!/bin/bash
+
+ros2 launch foxglove_bridge foxglove_bridge_launch.xml \
+    send_buffer_limit:=1000000000 num_threads:=4
```
Any reason for the limited num_threads? Have you seen a performance hit?

hect95 replied:
I left the argument to increase parallel execution. My reference is this quote from the Foxglove bridge doc

`
num_threads – Number of threads to use for the ROS node executor. This controls the number of subscriptions that can be processed in parallel. 0 means one thread per CPU core. Defaults to 0.
`

Up to some extend, the bigger the better? I guess I put 4 arbitrarily just to start somewhere. Have you tested this feature before?

GreatAlexander replied:
I have not tested it, I just left it to 0, but we should indeed play around with it since it could have been the reason we experienced issues with Foxglove viz.
Let's put it to 4 and see if we can play with the value in future.

GreatAlexander commented on scripts/container_tools/foxglove_bridge:
I think we need to add the `address:=172.31.0.1` or the bridge won't work as expected (Due to the multiple subnets in the compute)

hect95 replied:
O yeah, I forgot to put it back as I was doing some test locally.

GreatAlexander commented on scripts/container_tools/record_rosbag:
```diff
@@ -0,0 +1,43 @@
+OUTPUT_DIR="$ROS_WS/rosbags"
+
+# Generate a date and time prefix in the format YYYY_MM_DD-HH_MM_SS
+DATE_PREFIX=$(date "+%Y_%m_%d-%H_%M_%S")
```
It saddens me to see the reverse date order, but I guess it makes sense for alphabetic order...

hect95 replied:
That was my rationale, to distinguish older/newer bags, specially when recording and analysing. So are we Okay to leave it like that?

GreatAlexander replied:
Roger roger

GreatAlexander commented on scripts/host/upload_docker_image_to_vehicle.sh:
```diff
@@ -0,0 +1,72 @@
+
+# Use scp to transfer the tar file to the remote directory
+echo -e "${WHITE}Transferring ${CYAN}$FULL_TAR_PATH ${WHITE}to ${BLUE}$VEHICLE_USER@$VEHICLE_HOST:$REMOTE_DIR${NO_COLOR}"
+scp "$FULL_TAR_PATH" "${VEHICLE_USER}@${VEHICLE_HOST}:${REMOTE_DIR}"
```
Did you consider rsync after what we discussed? It could be safer and quicker (e.g. compressed + avoid transferring what's already there).

hect95 replied:
I have, I like that feature that only copies stuff that are actually different. I have change the cmd to do `rsync` instead

GreatAlexander commented on scripts/host/upload_docker_image_to_vehicle.sh:
Looks good, although as we discussed ideally we would pull tagged docker images into the compute from github, so this script would only be for development (e.g. build an unreleased image locally and run it on the compute), so perhaps a bit overboard atm.

hect95 replied:
Sorry, I didn't catch this one, could you elaborate? Maybe you mean the `runtime.sh` and the `dev.sh` instead?

GreatAlexander replied:
Nevermind, I just meant the script is quite advanced for what I expected, since hopefully we won't need to use this upload docker image script _that_ often (These days I think more than usual since we're working mostly on sensor containers and we want to test on the car asap).
The work is done already so no complains from me, was just a passing comment! :laughing:

hect95 committed:
Adding fixes as per PR comments:

- Using rsync instead of scp for data transfer
- Binding foxglove bridge to vehicle local ip
- Fixing vehicle details in `upload_docker_image_to_vehicle.sh`

hect95 committed:
Removing dots when mounting

hect95 committed:
Fixing dev mounting path issue

hect95 committed:
Removing dev/runtme tag for simplicity

GreatAlexander commented on scripts/host_tools/upload_docker_image_to_vehicle.sh:
```diff
@@ -0,0 +1,72 @@
+echo -e "${WHITE}Transferring ${CYAN}$FULL_TAR_PATH ${WHITE}to ${BLUE}$VEHICLE_USER@$VEHICLE_HOST:$REMOTE_DIR${NO_COLOR}"
+rsync "$FULL_TAR_PATH" "${VEHICLE_USER}@${VEHICLE_HOST}:${REMOTE_DIR}"
+
+# Checking if scp succeeded
```
```suggestion
# Check if rsync succeeded
```

hect95 replied:
Done

GreatAlexander replied:
Actually my comment had 2 changes, scp->rsync and checking->check (The rest of the script has imperative as the verb conjugation) which would then match the rest of the script.
Yesterday I was told off by pyflake in python for non-imperative descriptions so I think it's good for us to be consistent.

hect95 replied:
Done

hect95 committed:
Fixing typo

hect95 committed:
Fixing comment as per review

GreatAlexander commented on scripts/host_tools/install_vehicle_upload.sh:
```diff
@@ -0,0 +1,45 @@
+# #!/bin/bash
```
```suggestion
#!/bin/bash
```

hect95 replied:
Done

GreatAlexander commented on scripts/container_tools/record_rosbag:
Should this bash script and the foxglove_bridge script not end with a `.sh` file format?

hect95 replied:
I was thinking about it, but I though it  looked better without the `.sh` as are installed in the PATH of the contianer so you can access the cmd easily. We either remove the `.sh` of the other host scripts or add it for this two scripts, which one do you prefer?

GreatAlexander replied:
In general I like to have file extensions as part of the name to know what they are just by looking at their name, otherwise they could be anything (Typically no extension means a binary which shouldn't be human readable).

hect95 replied:
Okay so will put all of the bsah scripts with `.sh`

GreatAlexander commented on config/sensor_topics.txt:
Plz sort alphabetically
```
/cameras/fsp_l/camera_info
/cameras/fsp_l/image_raw
/cameras/fsp_l/meta
/cameras/lspf_r/camera_info
/cameras/lspf_r/image_raw
/cameras/lspf_r/meta
/cameras/lspr_l/camera_info
/cameras/lspr_l/image_raw
/cameras/lspr_l/meta
/cameras/rsp_l/camera_info
/cameras/rsp_l/image_raw
/cameras/rsp_l/meta
/cameras/rspf_l/camera_info
/cameras/rspf_l/image_raw
/cameras/rspf_l/meta
/cameras/rspr_r/camera_info
/cameras/rspr_r/image_raw
/cameras/rspr_r/meta
/diagnostics
/fix
/left_velodyne/velodyne_points
/points
/right_velodyne/velodyne_points
```

hect95 replied:
Adding more and now sorted in alphabetical order.

hect95 committed:
Adding .sh extension

hect95 committed:
Increasing max cache size for recording

hect95 committed:
Fix as per PR comments:

- Fixing typo on install_vehicle_upload.sh

- Renaming install_vehicle_upload.sh to
  install_tool.sh for simplicity

- Adding more sensor topics and sorting them
 in alphabetical order

hect95 committed:
Avoid recording leaf topics

hect95 committed:
Adding docker build workflow
